### PR TITLE
HIVE-21843: UNION query with regular expressions for column name does not work

### PR DESCRIPTION
**Pull request by kgyrtkirk**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 1]
```diff
@@ -4453,7 +4452,7 @@ private boolean isAggregateInSelect(Node node, Collection<ASTNode> aggregateFunc
    * Returns whether the pattern is a regex expression (instead of a normal
    * string). Normal string is a string with all alphabets/digits and "_".
    */
-  boolean isRegex(String pattern, HiveConf conf) {
+  static boolean isRegex(String pattern, HiveConf conf) {
```
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:4455]: This function is called from `SemanticAnalyzer` and its subclasses. I haven't found any other invocations. Is it necessary to change this to `static` ?

File: ql/src/test/org/apache/hadoop/hive/ql/parse/authorization/PrivilegesTestBase.java [hunk 2]
```diff
@@ -35,8 +37,9 @@
 
   public static void grantUserTable(String privStr, PrivilegeType privType, QueryState queryState, Hive db)
       throws Exception {
+    Context ctx=new Context(new HiveConf());
     DDLWork work = AuthorizationTestUtil.analyze(
-        "GRANT " + privStr + " ON TABLE " + TABLE + " TO USER " + USER, queryState, db);
+        "GRANT " + privStr + " ON TABLE " + TABLE + " TO USER " + USER, queryState, db,ctx);
```
Comment (kasakrisz) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/authorization/PrivilegesTestBase.java:42]: nit: space
`queryState, db, ctx);`

File: ql/src/test/org/apache/hadoop/hive/ql/parse/authorization/TestHiveAuthorizationTaskFactory.java [hunk 3]
```diff
@@ -457,7 +458,7 @@ public void testGrantServer() throws Exception {
   }
 
   private DDLWork analyze(String command) throws Exception {
-    return AuthorizationTestUtil.analyze(command, queryState, db);
+    return AuthorizationTestUtil.analyze(command, queryState, db,new Context(queryState.getConf()));
```
Comment (kasakrisz) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/authorization/TestHiveAuthorizationTaskFactory.java:461]: nit: space
`db, new Context`

File: ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java [hunk 4]
```diff
@@ -58,7 +76,7 @@ public void testSimpleQuery() {
     try {
       lep.getLineageInfo("INSERT OVERWRITE TABLE dest1 partition (ds = '111')  " 
           + "SELECT s.* FROM srcpart TABLESAMPLE (BUCKET 1 OUT OF 1) s " 
-          + "WHERE s.ds='2008-04-08' and s.hr='11'");
+          + "WHERE s.ds='2008-04-08' and s.hr='11'",ctx);
```
Comment (kasakrisz) [inline ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java:79]: nit: space
`s.hr='11'", ctx);`

File: ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java [hunk 5]
```diff
@@ -71,47 +89,37 @@ public void testSimpleQuery() {
...
+    i.add("src1");
+    checkOutput(lep, i, o);
   }
 
   @Test
   public void testSimpleQuery4() {
     LineageInfo lep = new LineageInfo();
     try {
-      lep.getLineageInfo("FROM ( FROM ( FROM src1 src1 SELECT src1.key AS c1, src1.value AS c2 WHERE src1.key > 10 and src1.key < 20) a RIGHT OUTER JOIN ( FROM src2 src2 SELECT src2.key AS c3, src2.value AS c4 WHERE src2.key > 15 and src2.key < 25) b ON (a.c1 = b.c3) SELECT a.c1 AS c1, a.c2 AS c2, b.c3 AS c3, b.c4 AS c4) c SELECT c.c1, c.c2, c.c3, c.c4");
+      lep.getLineageInfo("FROM ( FROM ( FROM src1 src1 SELECT src1.key AS c1, src1.value AS c2 WHERE src1.key > 10 and src1.key < 20) a RIGHT OUTER JOIN ( FROM src2 src2 SELECT src2.key AS c3, src2.value AS c4 WHERE src2.key > 15 and src2.key < 25) b ON (a.c1 = b.c3) SELECT a.c1 AS c1, a.c2 AS c2, b.c3 AS c3, b.c4 AS c4) c SELECT c.c1, c.c2, c.c3, c.c4",ctx);
```
Comment (kasakrisz) [inline ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java:122]: `c.c4", ctx);`

File: ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java [hunk 6]
```diff
@@ -128,7 +136,7 @@ public void testSimpleQuery5() {
     LineageInfo lep = new LineageInfo();
     try {
       lep.getLineageInfo("insert overwrite table x select a.y, b.y " 
-          + "from a a full outer join b b on (a.x = b.y)");
+          + "from a a full outer join b b on (a.x = b.y)",ctx);
```
Comment (kasakrisz) [inline ql/src/test/org/apache/hadoop/hive/ql/tool/TestLineageInfo.java:139]: `(a.x = b.y)", ctx);`

**Review by kasakrisz — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 1 above)
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:4455]: this patch adds a call to this method from `ParseUtils`  ; that's why it was neccessary to change it to at least `static`

Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:4455]: found it.

**Review by kasakrisz — APPROVED** [review]
+1 pending tests

**Pull request merged**